auth_service/auth_adapter/core/auth.py

```diff
@@ -18,121 +18,207 @@
+        else:
+            log.warning("No OIDC authority URL configured.")
+        client_id = config.oidc_client_id
+        if client_id:
+            self.check_claims["client_id"] = client_id
+        else:
+            log.warning("No OIDC client ID configured.")
+
+
+jwt_config = JWTConfig()
+
+
+def _compare_user(user: User, external_claims: dict[str, Any]) -> Optional[str]:
+    """Compare user with external claims and return inactivation context."""
+    if user.status == UserStatus.ACTIVATED:
+        if user.name != external_claims.get("name"):
+            return "name change"
+        if user.email != external_claims.get("email"):
+            return "email change"
+    return None
 
 
-def exchange_token(external_token: Optional[str]) -> Optional[str]:
+def _get_inactivated_user(user: User, context: str) -> User:
+    """Get an inactivated copy of the User object."""
+    return user.copy(
+        update=dict(
+            status=UserStatus.INACTIVATED.value,
+            status_change=StatusChange(
+                previous=user.status,
+                by=None,
+                context=context,
+                change_date=datetime.now(),
+            ),
+        )
+    )
+
+
+async def exchange_token(
+    external_token: Optional[str], pass_sub: bool = False, user_dao: UserDao = Depends()
+) -> Optional[str]:
     """Exchange the external token against an internal token.
 
     If the provided external token is valid, a corresponding internal token
     will be returned. Otherwise None will be returned.
+
+    The internal token will contain the name and email taken from the
+    external token, and also its issued date and expiry date.
+    If the user is already registered, the user id and status will be
+    included in the internal token as well.
+    If name or email do not match with the external token, the user
+    is automatically deactivated in the database and internal token.
+    If the user is not yet registered, and pass_sub is set, then the sub claim
+    will be included in the internal token as "ls_id", otherwise and empty string
+    will be returned instead of the internal token.
     """
-    payload = decode_and_verify_token(external_token)
-    if not payload or not isinstance(payload, dict):
-        return None
-    name = payload.get("name")
-    email = payload.get("email")
-    if not (name and email):
-        log.debug("Access token does not contain required user info")
-        return None
-    payload = dict(name=name, email=email)
-    internal_token = sign_and_encode_token(payload)
+    external_claims = decode_and_validate_token(external_token)
+    if not external_claims:
+        return None  # invalid token (error)
+    internal_claims = {
+        claim: external_claims[claim] for claim in jwt_config.copied_claims
+    }
+    sub = external_claims["sub"]
+    try:
+        user = await user_dao.find_one(mapping={"ls_id": sub})
+    except Exception as exc:  # pylint:disable=broad-except
```

Comment:
There are specific errors in hexkit that can be used here. However, if they occur we can also simply fail (let the service die), because this is a bug that we need to take care of and fix.

Reply:
Changed in 7372d21.